scipy/optimize/minpack.py

```diff
@@ -510,28 +523,42 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         values will all be 1 (if the number of parameters for the function
         can be determined using introspection, otherwise a ValueError
         is raised).
-    sigma : None or M-length sequence, optional
-        If not None, the uncertainties in the ydata array. These are used as
-        weights in the least-squares problem
-        i.e. minimising ``np.sum( ((f(xdata, *popt) - ydata) / sigma)**2 )``
-        If None, the uncertainties are assumed to be 1.
+    sigma : None or M-length sequence or MxM array, optional
+        Determines the uncertainty in `ydata`. If we define residuals as
+        ``e = ydata - f(xdata, *popt)``, then the interpretation of `sigma`
+        depends on its number of dimensions:
+
+            - A 1-d `sigma` should contain values of standard deviations of
+              uncertainties in `ydata`. In this case, the optimized function is
+              ``chisq = sum((e / sigma) ** 2)``.
+
+            - A 2-d `sigma` should contain the covariance matrix of
+              uncertainties in `ydata`. In this case, the optimized function is
+              ``chisq = e.T @ inv(sigma) @ e``.
+
+              .. versionadded:: 0.19
+
+        None (default) is equivalent of 1-d `sigma` filled with ones.
     absolute_sigma : bool, optional
-        If False, `sigma` denotes relative weights of the data points.
-        The returned covariance matrix `pcov` is based on *estimated*
-        errors in the data, and is not affected by the overall
-        magnitude of the values in `sigma`. Only the relative
-        magnitudes of the `sigma` values matter.
-
-        If True, `sigma` describes one standard deviation errors of
-        the input data points. The estimated covariance in `pcov` is
-        based on these values.
+        If True, `sigma` is used in an absolute sense and the estimated parameter
+        covariance `pcov` reflects these absolute values.
+
+        If False, `sigma` is used in a relative sense. The returned parameter
```

Comment:
Maybe merge the beginning and the ending to make it shorter overall: 

"If False, only the relative magnitudes  of the `sigma` values matter. ... `sigma` is scaled to match the sample variance of the residuals after the fit. Mathematically, ... "